icicle/appUtils/msm/msm.cu

```diff
@@ -636,66 +649,73 @@ namespace msm {
-          CHK_IF_RETURN(
-            cudaMallocAsync(&temp_buckets2, sizeof(P) * source_buckets_count / 2, stream)); // 32*2^8*2^7 buckets
+          CHK_IF_RETURN(cudaMallocAsync(
+            &target_buckets, sizeof(P) * target_buckets_count * batch_size, stream)); // 32*2^8*2^7 buckets
+          CHK_IF_RETURN(cudaMallocAsync(
+            &temp_buckets1, sizeof(P) * source_buckets_count / 2 * batch_size, stream)); // 32*2^8*2^7 buckets
+          CHK_IF_RETURN(cudaMallocAsync(
+            &temp_buckets2, sizeof(P) * source_buckets_count / 2 * batch_size, stream)); // 32*2^8*2^7 buckets
 
           if (source_bits_count > 0) {
             for (unsigned j = 0; j < target_bits_count; j++) {
-              // unsigned last_j = target_bits_count - 1;
-              unsigned nof_threads = (source_buckets_count >> (1 + j));
+              const bool is_first_iter = (j == 0);
+              const bool is_last_iter = (j == target_bits_count - 1);
+              unsigned nof_threads = (source_buckets_count >> (1 + j)) * batch_size;
               NUM_THREADS = min(MAX_TH, nof_threads);
               NUM_BLOCKS = (nof_threads + NUM_THREADS - 1) / NUM_THREADS;
               single_stage_multi_reduction_kernel<<<NUM_BLOCKS, NUM_THREADS, 0, stream>>>(
-                j == 0 ? source_buckets : temp_buckets1, j == target_bits_count - 1 ? target_buckets : temp_buckets1,
-                1 << (source_bits_count - j), j == target_bits_count - 1 ? 1 << target_bits_count : 0, 0, 0,
-                nof_threads);
+                is_first_iter ? source_buckets : temp_buckets1, is_last_iter ? target_buckets : temp_buckets1,
+                1 << (source_bits_count - j), is_last_iter ? 1 << target_bits_count : 0, 0 /*=write_phase*/,
+                0 /*=padding*/, nof_threads);
 
               NUM_THREADS = min(MAX_TH, nof_threads);
               NUM_BLOCKS = (nof_threads + NUM_THREADS - 1) / NUM_THREADS;
               single_stage_multi_reduction_kernel<<<NUM_BLOCKS, NUM_THREADS, 0, stream>>>(
-                j == 0 ? source_buckets : temp_buckets2, j == target_bits_count - 1 ? target_buckets : temp_buckets2,
-                1 << (target_bits_count - j), j == target_bits_count - 1 ? 1 << target_bits_count : 0, 1, 0,
-                nof_threads);
+                is_first_iter ? source_buckets : temp_buckets2, is_last_iter ? target_buckets : temp_buckets2,
+                1 << (target_bits_count - j), is_last_iter ? 1 << target_bits_count : 0, 1 /*=write_phase*/,
+                0 /*=padding*/, nof_threads);
             }
           }
           if (target_bits_count == 1) {
-            nof_bms = bitsize;
-            CHK_IF_RETURN(cudaMallocAsync(&final_results, sizeof(P) * nof_bms, stream));
+            // Note: the reduction ends up with 'target_windows_count' windows per batch element. for batch=1 we can
+            // only sum the first 'bitsize' elements to avoid a few ECADD which add zeros
```

Comment:
why isn't this true for every batch size?

Reply:
It is true for every batch size. 
The thing is that for fields like BLS12-377 where the scalar field is 253b we end up with >253 bms (e.g. 254 bms for c=2 with 1 iteration of reduction from 127 to 254 bms. For other cases can be a little more than that) where the most significant ones are guaranteed to be zero. 
The original code summed the bms based on bitsize so effectively summing only the non-zero bms. It is slightly more efficient more summing all bms (254 in the example I gave). I wrote the code this way for batch=1 since I did not want to end up with worse performance (although maybe negligible).

generally for batch>1 we must account for all bms so that the other batch elements are summing the right bms. If setting bm=253, the second batch element is summing wrong bms and all results end up wrong except for the first batch element.

Do you think I should always set bm based on the final number of windows? it is slightly less efficient but more drastic.

Reply:
why not have each msm sum only the non zero buckets? (and make sure that the buckets are the correct ones - I'm pretty sure you can use bm=253 and correct the kernel to sum the correct buckets) let's go for best performance

Reply:
Yes I think you are right. I will do that.